presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -317,12 +322,46 @@ private static Type getCommonSuperTypeForVarchar(VarcharType firstType, VarcharT
         return createVarcharType(Math.max(firstType.getLength(), secondType.getLength()));
     }
 
+    private Optional<Type> getCommonSuperTypeForRow(RowType firstType, RowType secondType)
+    {
+        ImmutableList.Builder<Type> commonParameterTypes = ImmutableList.builder();
+        ImmutableList.Builder<String> commonParameterNames = ImmutableList.builder();
+        List<RowField> firstFields = firstType.getFields();
+        List<RowField> secondFields = secondType.getFields();
+        if (firstFields.size() != secondFields.size()) {
+            return Optional.empty();
+        }
+
+        for (int i = 0; i < firstFields.size(); i++) {
+            Optional<Type> commonParameterType = getCommonSuperType(firstFields.get(i).getType(), secondFields.get(i).getType());
+            if (!commonParameterType.isPresent()) {
+                return Optional.empty();
+            }
+            commonParameterTypes.add(commonParameterType.get());
+
+            Optional<String> firstParameterName = firstFields.get(i).getName();
+            Optional<String> secondParameterName = secondFields.get(i).getName();
+            if (commonParameterNames != null
```

Comment:
If you use a `List<Optional<String>>` to hold the names, you the code can be simplified slightly:

```java
ImmutableList.Builder<Type> commonParameterTypes = ImmutableList.builder();
List<Optional<String>> commonParameterNames = new ArrayList<>();
for (int i = 0; i < firstFields.size(); i++) {
    Optional<Type> commonParameterType = getCommonSuperType(firstFields.get(i).getType(), secondFields.get(i).getType());
    if (!commonParameterType.isPresent()) {
        return Optional.empty();
    }
    commonParameterTypes.add(commonParameterType.get());

    Optional<String> firstParameterName = firstFields.get(i).getName();
    Optional<String> secondParameterName = secondFields.get(i).getName();
    if (firstParameterName.equals(secondParameterName)) {
        commonParameterNames.add(firstParameterName);
    }
    else {
        commonParameterNames.add(Optional.empty());
    }
}

List<String> names = null;
if (commonParameterNames.stream().allMatch(Optional::isPresent)) {
    names = commonParameterNames.stream()
            .map(Optional::get)
            .collect(toImmutableList());
}

return Optional.of(new RowType(commonParameterTypes.build(), Optional.ofNullable(names)));
```